cpp/src/utilities/stream_pool.cpp

```diff
@@ -0,0 +1,205 @@
+   *
+   * @param count The number of stream views to return.
+   * @return Vector containing `count` stream views.
+   */
+  virtual std::vector<rmm::cuda_stream_view> get_streams(uint32_t count) = 0;
+
+  /**
+   * @brief Get the number of stream objects in the pool.
+   *
+   * This function is thread safe with respect to other calls to the same function.
+   *
+   * @return the number of stream objects in the pool
+   */
+  virtual std::size_t get_stream_pool_size() const = 0;
+};
+
+/**
+ * @brief Implementation of `cuda_stream_pool` that wraps an `rmm::cuda_stram_pool`.
+ */
+class rmm_cuda_stream_pool : public cuda_stream_pool {
+  rmm::cuda_stream_pool _pool;
+
+ public:
+  rmm_cuda_stream_pool() : _pool{STREAM_POOL_SIZE} {}
+  rmm::cuda_stream_view get_stream() override { return _pool.get_stream(); }
+  rmm::cuda_stream_view get_stream(std::size_t stream_id) override
+  {
+    return _pool.get_stream(stream_id);
+  }
+
+  std::vector<rmm::cuda_stream_view> get_streams(uint32_t count) override
+  {
+    if (count > STREAM_POOL_SIZE) {
+      CUDF_LOG_WARN("get_streams called with count ({}) > pool size ({})", count, STREAM_POOL_SIZE);
+    }
+    auto streams = std::vector<rmm::cuda_stream_view>();
+    for (uint32_t i = 0; i < count; i++) {
+      streams.emplace_back(_pool.get_stream());
+    }
+    return streams;
+  }
+
+  std::size_t get_stream_pool_size() const override { return STREAM_POOL_SIZE; }
+};
+
+/**
+ * @brief Implementation of `cuda_stream_pool` that always returns `cudf::get_default_stream()`
+ */
+class debug_cuda_stream_pool : public cuda_stream_pool {
+ public:
+  rmm::cuda_stream_view get_stream() override { return cudf::get_default_stream(); }
+  rmm::cuda_stream_view get_stream(std::size_t stream_id) override
+  {
+    return cudf::get_default_stream();
+  }
+
+  std::vector<rmm::cuda_stream_view> get_streams(uint32_t count) override
+  {
+    return std::vector<rmm::cuda_stream_view>(count, cudf::get_default_stream());
+  }
+
+  std::size_t get_stream_pool_size() const override { return 1UL; }
+};
+
+/**
+ * @brief Initialize global stream pool.
+ */
+cuda_stream_pool* create_global_cuda_stream_pool()
+{
+  if (getenv("LIBCUDF_USE_DEBUG_STREAM_POOL")) return new debug_cuda_stream_pool();
+
+  return new rmm_cuda_stream_pool();
+}
+
+/**
+ * @brief RAII struct to wrap a cuda event and ensure it's proper destruction.
+ */
+struct cuda_event {
+  cuda_event()
+    : e_{[]() {
+        cudaEvent_t event;
+        CUDF_CUDA_TRY(cudaEventCreateWithFlags(&event, cudaEventDisableTiming));
+        return event;
+      }()}
+  {
+  }
+
+  operator cudaEvent_t() { return e_.get(); }
+
+ private:
+  struct deleter {
+    using pointer = cudaEvent_t;
+    auto operator()(cudaEvent_t e) { cudaEventDestroy(e); }
+  };
+  std::unique_ptr<cudaEvent_t, deleter> e_;
+};
+
+/**
+ * @brief Returns a cudaEvent_t for the current thread.
+ */
+cudaEvent_t event_for_thread()
+{
+  thread_local cuda_event thread_event;
```

Comment:
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**🚩 issue:** ‏ There's a tricky issue here. On multi-device systems, you will only have one event for all devices that this thread may use. Events are associated with a single device (the current device when `cudaEventCreate` is called. If this thread switches devices after `event_for_thread()` is called, and then it tries to `cudaEventRecord`, there will be a `cudaErrorInvalidResourceHandle`.  I'm currently working on an RMM PR to fix a similar issue, which I fixed by making `thread_local std::vector<cuda_event>` sized to the number of devices, instead of a single event. See https://github.com/rapidsai/rmm/pull/1333/files#diff-0d9e7691f84fe22b13d13977fe225d29f7824feee8e5b5763d98d8569b87634eR295

Reply:
@harrism I'm trying to implement something similar to your solution, but I'm worried about races between threads. Would it be appropriate to do something like:
```c++
std::vector<cuda_event> create_events_per_device()
{
  std::vector<cuda_event> result(get_num_cuda_devices());
  for (int i=0; i < result.size(); i++) {
    set_current_cuda_device(rmm::cuda_device_id{i});
    result[i] = cuda_event{};
  }
  return result;
}

cudaEvent_t event_for_thread()
{
  thread_local std::vector<cuda_event> thread_events = create_events_per_device();
  return thread_events[get_current_cuda_device().value()];
}
```

Reply:
I took a stab at duplicating what you are doing in `stream_ordered_memory_resource`, but now I'm wondering if we also need a stream pool per device?

Reply:
You definitely need either a stream pool per device or a stream pool that is multidevice aware. But the former can be done at the CUDF level outside of your stream pool class.  Even with that done, `thread_local` implies `static` so even if you create a pool per device your events need to be associated with a specific device.

Devices, streams and threads, Oh My!